climada/util/yearsets.py

```diff
@@ -22,223 +22,227 @@
-      """
+        rng = default_rng()
+        selected_events = rng.choice(indices, size=amount_events, replace=False,
+                                     p=probab_dis).astype('int')
+
+        if check_doubling: #check if an event occurs more than once in a year
+            unique_events = np.unique(selected_events)
+            #resample until each event occurs max. once per year
+            while len(unique_events) != len(selected_events):
+                rng = default_rng()
+                selected_events = rng.choice(indices, size=amount_events, replace=False,
+                                             p=probab_dis).astype('int')
+            check_doubling = False
+
+        idx_to_remove = [] #determine used events to remove them from sampling pool
+        for event in selected_events:
+            idx_to_remove.append(np.where(indices == event)[0][0])
+        indices = np.delete(indices, idx_to_remove)
+        freqs = np.delete(freqs, idx_to_remove)
+
+        #add the original indices and frequencies to the pool if there are less events
+        #in the pool than needed to fill the next sampled year
+        if (idx_year < (len(events_per_year)-1)) and (
+                len(indices) < events_per_year[idx_year+1]):
+            indices = np.append(indices, indices_orig)
+            freqs = np.append(freqs, freqs_orig)
+            check_doubling = True
 
-    impact_per_event = np.zeros(np.sum(sampling_dict['events_per_year']))
-    impact_per_year = np.zeros(len(sampling_dict['events_per_year']))
+        sampling_vect.append(selected_events)
 
-    for idx_event, event in enumerate(sampling_dict['selected_events']):
-        impact_per_event[idx_event] = event_impacts.at_event[sampling_dict[
-            'selected_events'][event]]
+    return sampling_vect
 
-    idx = 0
-    for year in range(len(sampling_dict['events_per_year'])):
-        impact_per_year[year] = np.sum(impact_per_event[idx:(idx+sampling_dict[
-            'events_per_year'][year])])
-        idx += sampling_dict['events_per_year'][year]
+def compute_imp_per_year(imp, sampling_vect):
+    """Sample annual impacts from the given event_impacts according to the sampling dictionary
+    Parameters:
+        imp : climada.engine.Impact()
+            impact object containing impacts per event
+        sampling_vect : 2D array
+            The sampling vector specifies how to sample the yimp, it consists of one
+            sub-array per sampled_year, which contains the event_ids of the events used to
+            calculate the annual impacts.
+    Returns:
+        imp_per_year: array
+            Sampled impact per year (length = sampled_years)
+      """
 
-    return impact_per_year
+    imp_per_year = np.zeros(len(sampling_vect))
+    for year, sampled_events in enumerate(sampling_vect):
+        if sampled_events.size > 0:
+            imp_per_year[year] = np.sum(imp.at_event[sampled_events])
```

Comment:
`np.sum([]) = 0.0`
```suggestion
    imp_per_year = [np.sum(imp.at_event[sampled_events]) for sampled_events in sampling_vect]
```

Reply:
The problem here is that in case of no event it's not np.sum([ ]) but np.sum(imp.at_event[ ]), which results in this error: IndexError: arrays used as indices must be of integer (or boolean) type. 
Do you have a suggestion on how to avoid this?

Reply:
This should solve the problem.
```suggestion
    imp_per_year = [np.sum(imp.at_event[list(sampled_events)]) for sampled_events in sampling_vect]
```

Reply:
Probably it doesn't make much sense to keep this as a separate function if it's literally one line, right?

Reply:
I do not have a strong opinion here. Please do as you prefer.